src/Lecture2.hs

```diff
@@ -285,4 +354,12 @@ Write a function that takes and expression and performs "Constant
 Folding" optimization on the given expression.
 -}
 constantFolding :: Expr -> Expr
-constantFolding = error "TODO"
+constantFolding expr =
+  case expr of 
+    Var x -> Var x
+    Lit x -> Lit x
+    Add (Lit 0) x -> x
+    Add x (Lit 0) -> x
+    Add (Lit x) (Lit y) -> Lit (x + y)
+    Add expr1 expr2 -> 
+      Add (constantFolding expr1) (constantFolding expr2)
```

Comment:
I don't have such test, but it looks like if I pass the following expression to this function:

```haskell
Add (Add (Var "x") (Lit 1)) (Add (Var "y") (Lit 3)
```

It won't fold all the constants since you fold both branches of `Add` independently 🤔 

I guess, I need to add more tests 😅

Reply:
Btw could you give a hint on how to fold nested expressions?